From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Wed, 18 Feb 2026 14:56:50 +0100
Subject: Move new symbol out of public library

As part of this patchset, we're backporting a new function
`sd_session_get_extra_device_access` which will be exported to libsystemd in
version 260. We don't want to export the symbol in out backport to version 258,
so move its definition (and one helper function) to its only user, udevd.

Forwarded: not-needed, downstream-only temporary patch
Bug-GNOME: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/224
Bug-Ubuntu: https://launchpad.net/bugs/2077538
---
 src/libsystemd/libsystemd.sym      |  1 -
 src/libsystemd/sd-login/sd-login.c | 30 -------------------
 src/systemd/sd-login.h             |  5 ----
 src/udev/udev-builtin-uaccess.c    | 60 ++++++++++++++++++++++++++++++++++++++
 4 files changed, 60 insertions(+), 36 deletions(-)

diff --git a/src/libsystemd/libsystemd.sym b/src/libsystemd/libsystemd.sym
index 439e872..a1fa794 100644
--- a/src/libsystemd/libsystemd.sym
+++ b/src/libsystemd/libsystemd.sym
@@ -1084,5 +1084,4 @@ global:
         sd_event_set_exit_on_idle;
         sd_event_get_exit_on_idle;
         sd_varlink_is_connected;
-        sd_session_get_extra_device_access;
 } LIBSYSTEMD_258;
diff --git a/src/libsystemd/sd-login/sd-login.c b/src/libsystemd/sd-login/sd-login.c
index 1331190..f65fb20 100644
--- a/src/libsystemd/sd-login/sd-login.c
+++ b/src/libsystemd/sd-login/sd-login.c
@@ -677,36 +677,6 @@ _public_ int sd_session_is_remote(const char *session) {
         return parse_boolean(s);
 }

-_public_ int sd_session_get_extra_device_access(const char *session, char ***ret_ids) {
-        _cleanup_free_ char *p = NULL, *s = NULL;
-        int r;
-
-        r = file_of_session(session, &p);
-        if (r < 0)
-                return r;
-
-        r = parse_env_file(/* f= */ NULL, p, "EXTRA_DEVICE_ACCESS", &s);
-        if (r == -ENOENT)
-                return -ENXIO;
-        if (r < 0)
-                return r;
-
-        _cleanup_strv_free_ char **ids = NULL;
-        size_t n_ids = 0;
-        if (!isempty(s)) {
-                ids = strv_split(s, /* separators= */ NULL);
-                if (!ids)
-                        return -ENOMEM;
-
-                n_ids = strv_length(ids);
-        }
-
-        if (ret_ids)
-                *ret_ids = TAKE_PTR(ids);
-
-        return n_ids;
-}
-
 _public_ int sd_session_get_state(const char *session, char **ret_state) {
         _cleanup_free_ char *p = NULL, *s = NULL;
         int r;
diff --git a/src/systemd/sd-login.h b/src/systemd/sd-login.h
index f9cfaf6..70f05cf 100644
--- a/src/systemd/sd-login.h
+++ b/src/systemd/sd-login.h
@@ -150,11 +150,6 @@ int sd_session_is_active(const char *session);
 /* Return 1 if the session is remote. */
 int sd_session_is_remote(const char *session);

-/* Return extra hardware devices that the session is granted access to.
- * For every $ID in the list, this adds access for all devices tagged with
- * "xaccess-$ID" in udev. */
-int sd_session_get_extra_device_access(const char *session, char ***ret_ids);
-
 /* Get state from session. Possible states: online, active, closing.
  * This function is a more generic version of sd_session_is_active(). */
 int sd_session_get_state(const char *session, char **ret_state);
diff --git a/src/udev/udev-builtin-uaccess.c b/src/udev/udev-builtin-uaccess.c
index 9fd006e..18b7f94 100644
--- a/src/udev/udev-builtin-uaccess.c
+++ b/src/udev/udev-builtin-uaccess.c
@@ -14,6 +14,66 @@
 #include "udev-builtin.h"
 #include "user-util.h"

+#include "path-util.h"
+#include "env-file.h"
+
+static int _file_of_session(const char *session, char **ret) {
+        char *p;
+        int r;
+
+        assert(ret);
+
+        if (session) {
+                if (!session_id_valid(session))
+                        return -EINVAL;
+
+                p = path_join("/run/systemd/sessions", session);
+        } else {
+                _cleanup_free_ char *buf = NULL;
+
+                r = sd_pid_get_session(0, &buf);
+                if (r < 0)
+                        return r;
+
+                p = path_join("/run/systemd/sessions", buf);
+        }
+        if (!p)
+                return -ENOMEM;
+
+        *ret = p;
+        return 0;
+}
+
+static int sd_session_get_extra_device_access(const char *session, char ***ret_ids) {
+        _cleanup_free_ char *p = NULL, *s = NULL;
+        int r;
+
+        r = _file_of_session(session, &p);
+        if (r < 0)
+                return r;
+
+        r = parse_env_file(/* f= */ NULL, p, "EXTRA_DEVICE_ACCESS", &s);
+        if (r == -ENOENT)
+                return -ENXIO;
+        if (r < 0)
+                return r;
+
+        _cleanup_strv_free_ char **ids = NULL;
+        size_t n_ids = 0;
+        if (!isempty(s)) {
+                ids = strv_split(s, /* separators= */ NULL);
+                if (!ids)
+                        return -ENOMEM;
+
+                n_ids = strv_length(ids);
+        }
+
+        if (ret_ids)
+                *ret_ids = TAKE_PTR(ids);
+
+        return n_ids;
+}
+
 static int builtin_uaccess(UdevEvent *event, int argc, char *argv[]) {
         sd_device *dev = ASSERT_PTR(ASSERT_PTR(event)->dev);
         _cleanup_strv_free_ char **sessions = NULL;
